File: bridges/go/bridge.go
package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"
	"time"
)

type SpawnMode string

const (
	SingleSession SpawnMode = "single-session"
	SameDir       SpawnMode = "same-dir"
	Worktree      SpawnMode = "worktree"
)

type SessionStatus string

const (
	Active      SessionStatus = "active"
	Completed   SessionStatus = "completed"
	Failed      SessionStatus = "failed"
	Interrupted SessionStatus = "interrupted"
)

type BridgeConfig struct {
	APIBaseURL         string
	SessionIngressURL  string
	Dir                string
	MaxSessions        int
	SpawnMode          SpawnMode
	SessionTimeoutMs   *int
	DebugFile          *string
	Verbose            bool
}

func NewBridgeConfig(apiBaseURL, sessionIngressURL, dir string, maxSessions int, spawnMode SpawnMode, sessionTimeoutMs *int, debugFile *string, verbose bool) *BridgeConfig {
	return &BridgeConfig{
		APIBaseURL:        apiBaseURL,
		SessionIngressURL: sessionIngressURL,
		Dir:               dir,
		MaxSessions:       maxSessions,
		SpawnMode:         spawnMode,
		SessionTimeoutMs:  sessionTimeoutMs,
		DebugFile:         debugFile,
		Verbose:           verbose,
	}
}

func CreateBridgeConfig(apiBaseURL, sessionIngressURL, dir string, maxSessions int, spawnMode SpawnMode) *BridgeConfig {
	return NewBridgeConfig(apiBaseURL, sessionIngressURL, dir, maxSessions, spawnMode, nil, nil, false)
}

type SessionHandle struct {
	SessionID       string
	Process         *exec.Cmd
	CurrentActivity map[string]interface{}
	Activities      []map[string]interface{}
	LastStderr      []string
	AccessToken     *string
}

func NewSessionHandle(sessionID string, process *exec.Cmd, currentActivity map[string]interface{}, activities []map[string]interface{}, lastStderr []string, accessToken *string) *SessionHandle {
	if currentActivity == nil {
		currentActivity = make(map[string]interface{})
	}
	if activities == nil {
		activities = []map[string]interface{}{}
	}
	if lastStderr == nil {
		lastStderr = []string{}
	}
	return &SessionHandle{
		SessionID:       sessionID,
		Process:         process,
		CurrentActivity: currentActivity,
		Activities:      activities,
		LastStderr:      lastStderr,
		AccessToken:     accessToken,
	}
}

func (sh *SessionHandle) UpdateAccessToken(token string) {
	sh.AccessToken = &token
}

func (sh *SessionHandle) Done() bool {
	if sh.Process == nil || sh.Process.ProcessState != nil {
		return true
	}
	return false
}

type SessionSpawnOpts struct {
	SessionID            string
	SDKURL               string
	AccessToken          string
	UseCCRv2             bool
	WorkerEpoch          *int
	OnFirstUserMessage   func(string)
}

func NewSessionSpawnOpts(sessionID, sdkURL, accessToken string, useCCRv2 bool, workerEpoch *int, onFirstUserMessage func(string)) *SessionSpawnOpts {
	return &SessionSpawnOpts{
		SessionID:          sessionID,
		SDKURL:             sdkURL,
		AccessToken:        accessToken,
		UseCCRv2:           useCCRv2,
		WorkerEpoch:        workerEpoch,
		OnFirstUserMessage: onFirstUserMessage,
	}
}

type SessionSpawner struct{}

func (ss *SessionSpawner) Spawn(opts *SessionSpawnOpts, dir string) (*SessionHandle, error) {
	useGoImplementation := true // Default to using Go implementation
	if useGoImplementation {
		return ss.spawnWithGo(opts)
	}
	return ss.spawnWithGo(opts) // Fallback to Go implementation
}

func (ss *SessionSpawner) spawnWithGo(opts *SessionSpawnOpts) (*SessionHandle, error) {
	goBridgePath := filepath.Join(filepath.Dir(os.Args[0]), "bridge")
	cmd := exec.Command(goBridgePath, "spawn", opts.SessionID, opts.SDKURL, opts.AccessToken)
	err := cmd.Start()
	if err != nil {
		fmt.Printf("[Bridge] Go implementation call failed: %v\n", err)
		
		rustResult, err := ss.trySpawnWithRust(opts)
		if err == nil && rustResult != nil {
			return rustResult, nil
		}
		
		tsResult, err := ss.trySpawnWithTypescript(opts)
		if err == nil && tsResult != nil {
			return tsResult, nil
		}
		
		cResult, err := ss.trySpawnWithC(opts)
		if err == nil && cResult != nil {
			return cResult, nil
		}
		
		return ss.spawnWithGo(opts)
	}
	
	return ss.createSessionHandleFromExternalResult(opts, "Go"), nil
}

func (ss *SessionSpawner) trySpawnWithRust(opts *SessionSpawnOpts) (*SessionHandle, error) {
	rustBridgePath := filepath.Join(filepath.Dir(os.Args[0]), "bridge_rust")
	if _, err := os.Stat(rustBridgePath); err == nil {
		fmt.Println("[Bridge] Using Rust implementation")
		cmd := exec.Command(rustBridgePath, "spawn", opts.SessionID, opts.SDKURL, opts.AccessToken)
		err := cmd.Start()
		if err != nil {
			fmt.Printf("[Bridge] Rust implementation execution failed: %v\n", err)
			return nil, err
		}
		return ss.createSessionHandleFromExternalResult(opts, "Rust"), nil
	}
	return nil, fmt.Errorf("rust bridge not found")
}

func (ss *SessionSpawner) trySpawnWithTypescript(opts *SessionSpawnOpts) (*SessionHandle, error) {
	tsBridgePath := filepath.Join(filepath.Dir(os.Args[0]), "bridge_ts")
	if _, err := os.Stat(tsBridgePath); err == nil {
		fmt.Println("[Bridge] Using TypeScript implementation")
		cmd := exec.Command("node", tsBridgePath, "spawn", opts.SessionID, opts.SDKURL, opts.AccessToken)
		err := cmd.Start()
		if err != nil {
			fmt.Printf("[Bridge] TypeScript implementation execution failed: %v\n", err)
			return nil, err
		}
		return ss.createSessionHandleFromExternalResult(opts, "TypeScript"), nil
	}
	return nil, fmt.Errorf("typescript bridge not found")
}

func (ss *SessionSpawner) trySpawnWithC(opts *SessionSpawnOpts) (*SessionHandle, error) {
	cBridgePath := filepath.Join(filepath.Dir(os.Args[0]), "bridge_c")
	if _, err := os.Stat(cBridgePath); err == nil {
		fmt.Println("[Bridge] Using C implementation")
		cmd := exec.Command(cBridgePath, "spawn", opts.SessionID, opts.SDKURL, opts.AccessToken)
		err := cmd.Start()
		if err != nil {
			fmt.Printf("[Bridge] C implementation execution failed: %v\n", err)
			return nil, err
		}
		return ss.createSessionHandleFromExternalResult(opts, "C"), nil
	}
	return nil, fmt.Errorf("c bridge not found")
}

func (ss *SessionSpawner) createSessionHandleFromExternalResult(opts *SessionSpawnOpts, implementation string) *SessionHandle {
	cmd := exec.Command("echo", fmt.Sprintf("session spawned by %s", implementation))
	cmd.Start()
	
	return NewSessionHandle(
		opts.SessionID,
		cmd,
		nil,
		nil,
		nil,
		&opts.AccessToken,
	)
}

func CreateSessionSpawner() *SessionSpawner {
	return &SessionSpawner{}
}

type BridgeAPIClient struct {
	BaseURL string
}

func NewBridgeAPIClient(baseURL string) *BridgeAPIClient {
	return &BridgeAPIClient{BaseURL: baseURL}
}

func (b *BridgeAPIClient) PollForWork(ctx context.Context, environmentID, environmentSecret string, signal chan os.Signal, reclaimOlderThanMs int) (map[string]interface{}, error) {
	return b.performPolling(ctx, environmentID, environmentSecret, signal, reclaimOlderThanMs)
}

func (b *BridgeAPIClient) performPolling(ctx context.Context, environmentID, environmentSecret string, signal chan os.Signal, reclaimOlderThanMs int) (map[string]interface{}, error) {
	return nil, nil
}

func (b *BridgeAPIClient) AcknowledgeWork(ctx context.Context, environmentID, workID, token string) error {
	return b.sendAcknowledgment(ctx, environmentID, workID, token)
}

func (b *BridgeAPIClient) sendAcknowledgment(ctx context.Context, environmentID, workID, token string) error {
	return nil
}

func (b *BridgeAPIClient) HeartbeatWork(ctx context.Context, environmentID, workID, token string) error {
	return b.sendHeartbeat(ctx, environmentID, workID, token)
}

func (b *BridgeAPIClient) sendHeartbeat(ctx context.Context, environmentID, workID, token string) error {
	return nil
}

func (b *BridgeAPIClient) StopWork(ctx context.Context, environmentID, workID string) error {
	return b.stopSpecificWork(ctx, environmentID, workID)
}

func (b *BridgeAPIClient) stopSpecificWork(ctx context.Context, environmentID, workID string) error {
	return nil
}

func (b *BridgeAPIClient) ReconnectSession(ctx context.Context, environmentID, sessionID string) error {
	return b.performReconnection(ctx, environmentID, sessionID)
}

func (b *BridgeAPIClient) performReconnection(ctx context.Context, environmentID, sessionID string) error {
	return nil
}

func (b *BridgeAPIClient) ArchiveSession(ctx context.Context, sessionID string) error {
	return b.archiveSpecificSession(ctx, sessionID)
}

func (b *BridgeAPIClient) archiveSpecificSession(ctx context.Context, sessionID string) error {
	return nil
}

func CreateBridgeAPIClient(baseURL string) *BridgeAPIClient {
	return NewBridgeAPIClient(baseURL)
}

type BridgeLogger struct {
	sessions       map[string]interface{}
	activeSessions int
	maxSessions    int
	spawnMode      SpawnMode
}

func NewBridgeLogger() *BridgeLogger {
	return &BridgeLogger{
		sessions:       make(map[string]interface{}),
		activeSessions: 0,
		maxSessions:    1,
		spawnMode:      SingleSession,
	}
}

func (bl *BridgeLogger) PrintBanner(config *BridgeConfig, environmentID string) {
	bannerMsg := fmt.Sprintf("Bridge started with environment: %s", environmentID)
	fmt.Println(bannerMsg)
}

func (bl *BridgeLogger) LogSessionStart(sessionID, description string) {
	startMsg := fmt.Sprintf("Session started: %s - %s", sessionID, description)
	fmt.Println(startMsg)
}

func (bl *BridgeLogger) LogSessionComplete(sessionID string, durationMs int) {
	completeMsg := fmt.Sprintf("Session completed: %s - %dms", sessionID, durationMs)
	fmt.Println(completeMsg)
}

func (bl *BridgeLogger) LogSessionFailed(sessionID, message string) {
	failMsg := fmt.Sprintf("Session failed: %s - %s", sessionID, message)
	fmt.Println(failMsg)
}

func (bl *BridgeLogger) LogError(message string) {
	errorMsg := fmt.Sprintf("Error: %s", message)
	fmt.Println(errorMsg)
}

func (bl *BridgeLogger) LogVerbose(message string) {
	verboseMsg := fmt.Sprintf("Verbose: %s", message)
	fmt.Println(verboseMsg)
}

func (bl *BridgeLogger) LogReconnected(disconnectedMs int) {
	reconnMsg := fmt.Sprintf("Reconnected after %dms", disconnectedMs)
	fmt.Println(reconnMsg)
}

func (bl *BridgeLogger) SetAttached(sessionID string) {
}

func (bl *BridgeLogger) UpdateSessionCount(count, maxCount int, spawnMode SpawnMode) {
	bl.activeSessions = count
	bl.maxSessions = maxCount
	bl.spawnMode = spawnMode
}

func (bl *BridgeLogger) UpdateIdleStatus() {
}

func (bl *BridgeLogger) UpdateSessionStatus(sessionID, elapsed string, activity map[string]interface{}, trail []string) {
}

func (bl *BridgeLogger) AddSession(sessionID, url string) {
}

func (bl *BridgeLogger) RemoveSession(sessionID string) {
}

func (bl *BridgeLogger) SetSessionTitle(sessionID, title string) {
}

func (bl *BridgeLogger) UpdateSessionActivity(sessionID string, activity map[string]interface{}) {
}

func (bl *BridgeLogger) RefreshDisplay() {
}

func (bl *BridgeLogger) ClearStatus() {
}

func (bl *BridgeLogger) SetDebugLogPath(path string) {
}

func CreateBridgeLogger() *BridgeLogger {
	return NewBridgeLogger()
}

func RunBridgeLoop(
	ctx context.Context,
	config *BridgeConfig,
	environmentID, environmentSecret string,
	api *BridgeAPIClient,
	spawner *SessionSpawner,
	logger *BridgeLogger,
	signal chan os.Signal,
	initialSessionID *string,
) error {
	err := initializeBridge(config, environmentID, logger)
	if err != nil {
		return err
	}
	
	return startMainLoop(ctx, config, environmentID, environmentSecret, api, spawner, logger, signal, initialSessionID)
}

func initializeBridge(config *BridgeConfig, environmentID string, logger *BridgeLogger) error {
	logger.PrintBanner(config, environmentID)
	return nil
}

func startMainLoop(
	ctx context.Context,
	config *BridgeConfig,
	environmentID, environmentSecret string,
	api *BridgeAPIClient,
	spawner *SessionSpawner,
	logger *BridgeLogger,
	signal chan os.Signal,
	initialSessionID *string,
) error {
	fmt.Println("Bridge loop started")
	
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case sig := <-signal:
			fmt.Printf("Received signal: %s\n", sig)
			return nil
		default:
			time.Sleep(100 * time.Millisecond)
		}
	}
}

type BridgeClient = BridgeAPIClient

func main() {
	config := CreateBridgeConfig("http://example.com", "http://ingress.example.com", "/tmp", 1, SingleSession)
	
	sessionSpawner := CreateSessionSpawner()
	
	logger := CreateBridgeLogger()
	
	apiClient := CreateBridgeAPIClient("http://api.example.com")
	
	signalChan := make(chan os.Signal, 1)
	
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	
	initialSessionID := "test-session-id"
	
	err := RunBridgeLoop(ctx, config, "env-123", "secret-456", apiClient, sessionSpawner, logger, signalChan, &initialSessionID)
	if err != nil {
		fmt.Printf("Bridge loop error: %v\n", err)
	}
}

File: bridges/rust/bridge.rs
use std::collections::HashMap;
use std::process::{Child, Command};
use std::sync::Arc;
use tokio::sync::Mutex;

pub enum SpawnMode {
    SingleSession,
    SameDir,
    Worktree,
}

impl std::fmt::Display for SpawnMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SpawnMode::SingleSession => write!(f, "single-session"),
            SpawnMode::SameDir => write!(f, "same-dir"),
            SpawnMode::Worktree => write!(f, "worktree"),
        }
    }
}

pub enum SessionStatus {
    Active,
    Completed,
    Failed,
    Interrupted,
}

impl std::fmt::Display for SessionStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SessionStatus::Active => write!(f, "active"),
            SessionStatus::Completed => write!(f, "completed"),
            SessionStatus::Failed => write!(f, "failed"),
            SessionStatus::Interrupted => write!(f, "interrupted"),
        }
    }
}

pub struct BridgeConfig {
    pub api_base_url: String,
    pub session_ingress_url: String,
    pub dir: String,
    pub max_sessions: usize,
    pub spawn_mode: SpawnMode,
    pub session_timeout_ms: Option<u64>,
    pub debug_file: Option<String>,
    pub verbose: bool,
}

impl BridgeConfig {
    pub fn new(
        api_base_url: String,
        session_ingress_url: String,
        dir: String,
        max_sessions: usize,
        spawn_mode: SpawnMode,
        session_timeout_ms: Option<u64>,
        debug_file: Option<String>,
        verbose: bool,
    ) -> Self {
        Self {
            api_base_url,
            session_ingress_url,
            dir,
            max_sessions,
            spawn_mode,
            session_timeout_ms,
            debug_file,
            verbose,
        }
    }
}

pub fn create_bridge_config(
    api_base_url: String,
    session_ingress_url: String,
    dir: String,
    max_sessions: usize,
    spawn_mode: SpawnMode,
) -> BridgeConfig {
    BridgeConfig::new(
        api_base_url,
        session_ingress_url,
        dir,
        max_sessions,
        spawn_mode,
        Some(3600000), // 默认超时时间
        None,
        false,
    )
}

pub struct SessionHandle {
    pub session_id: String,
    pub process: Child,
    pub current_activity: HashMap<String, serde_json::Value>,
    pub activities: Vec<HashMap<String, serde_json::Value>>,
    pub last_stderr: Vec<String>,
    pub access_token: Option<String>,
}

impl SessionHandle {
    pub fn new(
        session_id: String,
        process: Child,
        current_activity: Option<HashMap<String, serde_json::Value>>,
        activities: Option<Vec<HashMap<String, serde_json::Value>>>,
        last_stderr: Option<Vec<String>>,
        access_token: Option<String>,
    ) -> Self {
        Self {
            session_id,
            process,
            current_activity: current_activity.unwrap_or_default(),
            activities: activities.unwrap_or_default(),
            last_stderr: last_stderr.unwrap_or_default(),
            access_token,
        }
    }

    pub fn update_access_token(&mut self, token: String) {
        self.access_token = Some(token);
    }

    pub fn done(&self) -> bool {
        self.process.try_wait().unwrap().is_some()
    }
}

pub type OnFirstUserMessageCallback = Box<dyn Fn(String) + Send>;

pub struct SessionSpawnOpts {
    pub session_id: String,
    pub sdk_url: String,
    pub access_token: String,
    pub use_ccr_v2: bool,
    pub worker_epoch: Option<i64>,
    pub on_first_user_message: Option<OnFirstUserMessageCallback>,
}

impl SessionSpawnOpts {
    pub fn new(
        session_id: String,
        sdk_url: String,
        access_token: String,
        use_ccr_v2: bool,
        worker_epoch: Option<i64>,
        on_first_user_message: Option<OnFirstUserMessageCallback>,
    ) -> Self {
        Self {
            session_id,
            sdk_url,
            access_token,
            use_ccr_v2,
            worker_epoch,
            on_first_user_message,
        }
    }
}

pub struct SessionSpawner;

impl SessionSpawner {
    pub fn new() -> Self {
        Self
    }

    pub fn spawn(&self, opts: SessionSpawnOpts, dir: &str) -> SessionHandle {
        let process = self.create_session_process();
        SessionHandle::new(
            opts.session_id,
            process,
            None,
            None,
            None,
            Some(opts.access_token),
        )
    }

    fn create_session_process(&self) -> Child {
        Command::new("echo")
            .arg("session spawned")
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn()
            .expect("Failed to start session process")
    }
}

pub fn create_session_spawner() -> SessionSpawner {
    SessionSpawner::new()
}

pub struct BridgeApiClient {
    pub base_url: String,
}

impl BridgeApiClient {
    pub fn new(base_url: String) -> Self {
        Self { base_url }
    }

    pub async fn poll_for_work(
        &self,
        environment_id: &str,
        environment_secret: &str,
        signal: Option<tokio::sync::broadcast::Receiver<()>>,
        reclaim_older_than_ms: i64,
    ) -> Option<HashMap<String, serde_json::Value>> {
        self.perform_polling(environment_id, environment_secret, signal, reclaim_older_than_ms)
            .await
    }

    async fn perform_polling(
        &self,
        _environment_id: &str,
        _environment_secret: &str,
        _signal: Option<tokio::sync::broadcast::Receiver<()>>,
        _reclaim_older_than_ms: i64,
    ) -> Option<HashMap<String, serde_json::Value>> {
        None
    }

    pub async fn acknowledge_work(
        &self,
        environment_id: &str,
        work_id: &str,
        token: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.send_acknowledgment(environment_id, work_id, token).await
    }

    async fn send_acknowledgment(
        &self,
        _environment_id: &str,
        _work_id: &str,
        _token: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        Ok(())
    }

    pub async fn heartbeat_work(
        &self,
        environment_id: &str,
        work_id: &str,
        token: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.send_heartbeat(environment_id, work_id, token).await
    }

    async fn send_heartbeat(
        &self,
        _environment_id: &str,
        _work_id: &str,
        _token: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        Ok(())
    }

    pub async fn stop_work(
        &self,
        environment_id: &str,
        work_id: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.stop_specific_work(environment_id, work_id).await
    }

    async fn stop_specific_work(
        &self,
        _environment_id: &str,
        _work_id: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        Ok(())
    }

    pub async fn reconnect_session(
        &self,
        environment_id: &str,
        session_id: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.perform_reconnection(environment_id, session_id).await
    }

    async fn perform_reconnection(
        &self,
        _environment_id: &str,
        _session_id: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        Ok(())
    }

    pub async fn archive_session(
        &self,
        session_id: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        self.archive_specific_session(session_id).await
    }

    async fn archive_specific_session(
        &self,
        _session_id: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        Ok(())
    }
}

pub fn create_bridge_api_client(base_url: String) -> BridgeApiClient {
    BridgeApiClient::new(base_url)
}

pub struct BridgeLogger {
    sessions: Arc<Mutex<HashMap<String, String>>>,
    active_sessions: usize,
    max_sessions: usize,
    spawn_mode: SpawnMode,
}

impl BridgeLogger {
    pub fn new() -> Self {
        Self {
            sessions: Arc::new(Mutex::new(HashMap::new())),
            active_sessions: 0,
            max_sessions: 1,
            spawn_mode: SpawnMode::SingleSession,
        }
    }

    pub fn print_banner(&self, config: &BridgeConfig, environment_id: &str) {
        let banner_msg = format!("Bridge started with environment: {}", environment_id);
        println!("{}", banner_msg);
    }

    pub fn log_session_start(&self, session_id: &str, description: &str) {
        let start_msg = format!("Session started: {} - {}", session_id, description);
        println!("{}", start_msg);
    }

    pub fn log_session_complete(&self, session_id: &str, duration_ms: u64) {
        let complete_msg = format!("Session completed: {} - {}ms", session_id, duration_ms);
        println!("{}", complete_msg);
    }

    pub fn log_session_failed(&self, session_id: &str, message: &str) {
        let fail_msg = format!("Session failed: {} - {}", session_id, message);
        println!("{}", fail_msg);
    }

    pub fn log_error(&self, message: &str) {
        let error_msg = format!("Error: {}", message);
        println!("{}", error_msg);
    }

    pub fn log_verbose(&self, message: &str) {
        let verbose_msg = format!("Verbose: {}", message);
        println!("{}", verbose_msg);
    }

    pub fn log_reconnected(&self, disconnected_ms: u64) {
        let reconn_msg = format!("Reconnected after {}ms", disconnected_ms);
        println!("{}", reconn_msg);
    }

    pub fn set_attached(&self, session_id: &str) {
    }

    pub fn update_session_count(&mut self, count: usize, max_count: usize, spawn_mode: SpawnMode) {
        self.active_sessions = count;
        self.max_sessions = max_count;
        self.spawn_mode = spawn_mode;
    }

    pub fn update_idle_status(&self) {
    }

    pub fn update_session_status(
        &self,
        session_id: &str,
        elapsed: &str,
        activity: &HashMap<String, serde_json::Value>,
        trail: &[String],
    ) {
    }

    pub async fn add_session(&self, session_id: String, url: String) {
        let mut sessions = self.sessions.lock().await;
        sessions.insert(session_id, url);
    }

    pub async fn remove_session(&self, session_id: &str) {
        let mut sessions = self.sessions.lock().await;
        sessions.remove(session_id);
    }

    pub fn set_session_title(&self, session_id: &str, title: &str) {
    }

    pub fn update_session_activity(
        &self,
        session_id: &str,
        activity: &HashMap<String, serde_json::Value>,
    ) {
    }

    pub fn refresh_display(&self) {
    }

    pub fn clear_status(&self) {
    }

    pub fn set_debug_log_path(&self, path: &str) {
    }
}

pub fn create_bridge_logger() -> BridgeLogger {
    BridgeLogger::new()
}

pub async fn run_bridge_loop(
    config: BridgeConfig,
    environment_id: String,
    environment_secret: String,
    api: BridgeApiClient,
    spawner: SessionSpawner,
    logger: BridgeLogger,
    signal: Option<tokio::sync::broadcast::Receiver<()>>,
    initial_session_id: Option<String>,
) {
    initialize_bridge(&config, &environment_id, &logger).await;
    start_main_loop(
        config,
        environment_id,
        environment_secret,
        api,
        spawner,
        logger,
        signal,
        initial_session_id,
    )
    .await;
}

async fn initialize_bridge(config: &BridgeConfig, environment_id: &str, logger: &BridgeLogger) {
    logger.print_banner(config, environment_id);
}

async fn start_main_loop(
    config: BridgeConfig,
    environment_id: String,
    environment_secret: String,
    api: BridgeApiClient,
    spawner: SessionSpawner,
    logger: BridgeLogger,
    signal: Option<tokio::sync::broadcast::Receiver<()>>,
    initial_session_id: Option<String>,
) {
    println!("Bridge loop started");
}

pub type BridgeClient = BridgeApiClient;

mod tests {
    use super::*;

    fn test_spawn_mode_display() {
        assert_eq!(format!("{}", SpawnMode::SingleSession), "single-session");
        assert_eq!(format!("{}", SpawnMode::SameDir), "same-dir");
        assert_eq!(format!("{}", SpawnMode::Worktree), "worktree");
    }

    fn test_session_status_display() {
        assert_eq!(format!("{}", SessionStatus::Active), "active");
        assert_eq!(format!("{}", SessionStatus::Completed), "completed");
        assert_eq!(format!("{}", SessionStatus::Failed), "failed");
        assert_eq!(format!("{}", SessionStatus::Interrupted), "interrupted");
    }
}